cmd/internal/kube/injector/injector.go

```diff
@@ -0,0 +1,204 @@
+		// unstructured objects because we likely won't know the type of all objects
+		// ahead of time (e.g., when reading multiple objects from a YAML file).
+		objects []*unstructured.Unstructured
+	}
+)
+
+// Constructs a new Injector with Kubernetes objects derived from the given file path.
+func FromYAML(filePath string) (Injector, error) {
+	yamlContent, err := getFile(filePath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Read the YAML file into a list of unstructured objects.
+	// This is necessary because the YAML file may contain multiple Kubernetes objects.
+	// We only want to inject the sidecar into Deployment objects, but we still need to parse all resources.
+	multidocReader := kyamlutil.NewYAMLReader(bufio.NewReader(bytes.NewReader(yamlContent)))
+
+	var objList []*unstructured.Unstructured
+	for {
+		raw, err := multidocReader.Read()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return nil, err
+		}
+
+		obj, err := fromRawObject(raw)
+		if err != nil {
+			return nil, err
+		}
+
+		result, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+		if err != nil {
+			return nil, err
+		}
+
+		objList = append(objList, &unstructured.Unstructured{Object: result})
+	}
+
+	return &injectorImpl{objects: objList}, nil
+}
+
+func (i *injectorImpl) InjectableNamespaces() ([]string, error) {
+	set := make(map[string]struct{})
+	for _, obj := range i.objects {
+		gvk := obj.GetObjectKind().GroupVersionKind()
+
+		if !isInjectable(gvk) {
+			continue
+		}
+
+		var deployment *appsv1.Deployment
+		err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.Object, &deployment)
+		if err != nil {
+			return nil, err
+		}
+
+		if deployment.Namespace == "" {
+			set["default"] = struct{}{}
+		} else {
+			set[deployment.Namespace] = struct{}{}
+		}
+	}
+
+	return sets.StringKeySet(set).List(), nil
+}
+
+func (i *injectorImpl) Inject(sidecar v1.Container) ([]*unstructured.Unstructured, error) {
+	onMap := func(obj *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+		out := obj.DeepCopy()
+		gvk := out.GetObjectKind().GroupVersionKind()
+
+		if !isInjectable(gvk) {
+			return out, nil
+		}
+
+		var deployment *appsv1.Deployment
+		err := runtime.DefaultUnstructuredConverter.FromUnstructured(out.Object, &deployment)
+		if err != nil {
+			return nil, err
+		}
+
+		injectedDeployment := deployment.DeepCopy()
+		if err != nil {
+			return nil, err
+		}
+
+		containers := injectedDeployment.Spec.Template.Spec.Containers
+		injectedDeployment.Spec.Template.Spec.Containers = append(containers, sidecar)
+
+		if err != nil {
+			return nil, err
+		}
+
+		out.Object, err = runtime.DefaultUnstructuredConverter.ToUnstructured(injectedDeployment)
+		if err != nil {
+			return nil, err
+		}
+
+		return out, nil
+	}
+
+	return mapUnstructured(i.objects, onMap)
+}
+
+func isInjectable(kind schema.GroupVersionKind) bool {
+	return kind.Group == "apps" && kind.Version == "v1" && kind.Kind == "Deployment"
```

Comment:
Would it work to define a schema.GroupVersionKind as a variable and compare with ==?  This is not a big difference (in fact it would be more lines of code) but you could then omit this helper entirely and do == or switch comparison at the point of use.

Reply:
I kind of like the idea of keeping the helper function since it's already being used by two methods (`InjectableNamespaces` and `Inject`). However, I have simplified the check as you've suggested in https://github.com/akitasoftware/akita-cli/pull/207/commits/12f88e2d4b2404332e9551064422d5be0ab6dfc8